yarn-project/end-to-end/src/e2e_fees/private_refunds.test.ts

```diff
@@ -0,0 +1,178 @@
+import { type PrivateFPCContract, PrivateTokenContract } from '@aztec/noir-contracts.js';
+
+import { expectMapping } from '../fixtures/utils.js';
+import { FeesTest } from './fees_test.js';
+
+describe('e2e_fees/private_refunds', () => {
+  let aliceWallet: Wallet;
+  let aliceAddress: AztecAddress;
+  let privateToken: PrivateTokenContract;
+  let privateFPC: PrivateFPCContract;
+
+  let InitialAlicePrivateTokens: bigint;
+  let InitialBobPrivateTokens: bigint;
+  let InitialPrivateFPCGas: bigint;
+
+  const t = new FeesTest('private_refunds');
+
+  beforeAll(async () => {
+    await t.applyInitialAccountsSnapshot();
+    await t.applyPublicDeployAccountsSnapshot();
+    await t.applyDeployGasTokenSnapshot();
+    await t.applyPrivateTokenAndFPC();
+    await t.applyFundAliceWithPrivateTokens();
+    ({ aliceWallet, aliceAddress, privateFPC, privateToken } = await t.setup());
+    t.logger.debug(`Alice address: ${aliceAddress}`);
+  });
+
+  afterAll(async () => {
+    await t.teardown();
+  });
+
+  beforeEach(async () => {
+    [[InitialAlicePrivateTokens, InitialBobPrivateTokens], [InitialPrivateFPCGas]] = await Promise.all([
+      t.privateTokenBalances(aliceAddress, t.bobAddress),
+      t.gasBalances(privateFPC.address),
+    ]);
+  });
+
+  it('can do private payments and refunds', async () => {
+    const bobKeyHash = t.bobWallet.getCompleteAddress().publicKeys.masterNullifierPublicKey.hash();
+    const rebateNonce = new Fr(42);
+    const tx = await privateToken.methods
+      .private_get_name()
+      .send({
+        fee: {
+          gasSettings: t.gasSettings,
+          paymentMethod: new PrivateRefundPaymentMethod(
+            privateToken.address,
+            privateFPC.address,
+            aliceWallet,
+            rebateNonce,
+            bobKeyHash,
+          ),
+        },
+      })
+      .wait();
+
+    expect(tx.transactionFee).toBeGreaterThan(0);
+
+    const refundedNoteValue = t.gasSettings.getFeeLimit().sub(new Fr(tx.transactionFee!));
+    const aliceKeyHash = t.aliceWallet.getCompleteAddress().publicKeys.masterNullifierPublicKey.hash();
+    const aliceRefundNote = new Note([refundedNoteValue, aliceKeyHash, rebateNonce]);
+    await t.aliceWallet.addNote(
+      new ExtendedNote(
+        aliceRefundNote,
+        t.aliceAddress,
+        privateToken.address,
+        PrivateTokenContract.storage.balances.slot,
+        PrivateTokenContract.notes.TokenNote.id,
+        tx.txHash,
+      ),
+    );
+
+    const bobFeeNote = new Note([new Fr(tx.transactionFee!), bobKeyHash, rebateNonce]);
+    await t.bobWallet.addNote(
+      new ExtendedNote(
+        bobFeeNote,
+        t.bobAddress,
+        privateToken.address,
+        PrivateTokenContract.storage.balances.slot,
+        PrivateTokenContract.notes.TokenNote.id,
+        tx.txHash,
+      ),
+    );
+
+    await expectMapping(t.gasBalances, [privateFPC.address], [InitialPrivateFPCGas - tx.transactionFee!]);
```

Comment:
I might be retarded but I feel like functions like expectMapping and then having complex types like `export type BalancesFn = ReturnType<typeof getBalancesFn>;` in tests just makes it all hard to read.